### ISSUE TITLE
Create script for running test/cql tests against Cassandra

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#10098]
The _approval tests_ in `test/cql` are simple files of CQL requests (the `*.cql` files) and their expected responses (`*.result`).

The test.py script executes the CQL requests not through `cqlsh` but through a special modified version of Scylla called `cql_repl`. Which works great for testing Scylla, but does not allow testing whether the tests pass identically against Cassandra. Running the same tests against Cassandra will allow us to check that not only does Scylla's behavior not regress across versions (this is so-called _approval testing_) but also that its current state it is compatible with Cassandra and we're not just enshrining some wrong behavior.

Yes, a developer can manually run Cassandra, run cqlsh on each `*.cql`, and check if the output matches the corresponding `*.result` file. But doing this is manually rather tedious, and therefore not even expected of test/cql test developers :-( So what I'm asking in this issue is to write a script, analogous to `test/cql-pytest/run-cassandra` - which does all of the above automatically: It starts Cassandra in a temporary directory, sends it CQL commands from `*.cql` file (can do that from the Python driver, for example), and compares the results - probably not a direct string comparison but something which ignores whitespaces and perhaps other types of known non-interesting differences between the Scylla and Cassandra output formats.

The script can be more than analogous to `test/cql-pytest/run-cassandra` - it can probably be based on it (and the underlying `test/cql-pytest/run.py` library).

CC @kostja

**Comment by nyh** [comment]
Recently @kostja modified test.py to run test/cql not using the separate cql_repl executable (which was dropped) and rather run Scylla and then use a small function program to run the CQL requests against it.

So today, It should be fairly straightforward to write a script similar to test/cql-ptest/run-cassandra which runs Cassandra and then use the same Python function that we already have to run the CQL requests against it and compare the result files.

An open question is whether all tests will pass on Cassandra this way, or we'll expect to see silly formatting differences on one hand, and deliberate differences on the other hand. We will need to solve the problem of silly differences, but discovering (and documenting) deliberate differences is an important benefit of such tests. See the "cassandra-bug", "scylla-only" tests is test/cql-pytest with cases where the differences are known, deliberate, and documented.

**Comment by kostja** [comment]
I think we are mixing up concepts here. test.py framework purpose is regression testing. If it does more than one thing it does them badly. Once again I encourage to look at this from the testing pyramid point of view: test.py purpose is to build a great foundation of the testing pyramid.

Once we bring Cassandra into the picture, the context becomes more broad. Why Cassandra only? Maybe we should cross-test Scylla 5 and Scylla 4? And also, what is the workflow for a failure? 

I think we're lacking test failure workflows very badly.

Is running against Cassandra part of next-master promotion? Is it part of release promotion? What is the procedure for flaky or broken tests? 

My concern is that we're blurring the focus of test.py. It would be nice to make it a good regression testing tool then anything else.

**Comment by nyh** [comment]
@kostja the purpose of being able to run tests against Cassandra is **not** to find bugs or regressions in Scylla. Rather the goal is to find bugs in the **functional tests** themselves. Usually while writing those tests:

In more than one case in the past, we had an incorrect functional test, which was written by a developer who made a mistake in developing a feature, and then repeated the same mistake in the test. With cql-repl, you don't even need to "repeat" the same mistake in the test - the test simply "enshrines" the incorrect output and deems it the gold standard.

My claim is that it doesn't have to be this way: We can actually run the cql-repl tests against Cassandra as well. I'm not suggesting that this should happen automatically in Jenkins or "next promotion" or anything like that. I'm just suggesting that a person who develops tests can check the validity of their test in this way. It can only be checked once on today's Cassandra, or perhaps can be repeated once a year on newer Cassandras just to learn if something changed in Cassandra without us realizing. I think this sort of philosophy has been very productive for test/cql-pytest, test/alternator and test/redis, and there is no reason why it can't be used for cql-repl tests as well.

Clearly, I'm proposing this mainly because 99% of Scylla's functional features are meant to be identical or almost identical to the ones of Cassandra (or DynamoDB or Redis). If this weren't the case, and 90% of Scylla's functionality was own invention with no parallel in Cassandra, I would have never proposed to run our tests on Cassandra.

Whether running-on-Cassandra should be part of the one-big-script "test.py" is a separate question. As you know, I'm not a big fan of monolithic scripts. In my older set of scripts, test/cql-pytest/run and test/cql-pytest/run-cassandra are two separate scripts (which share a lot of common code), the first was run in Jenkins, the second wasn't. So no, I don't think it must be part of test.py to be useful.

Finally, you asked "Once we bring Cassandra into the picture, the context becomes more broad. Why Cassandra only? Maybe we should cross-test Scylla 5 and Scylla 4?". This is an excellent question, which I've already answered in the past:
1. It's already not Cassandra-only. We run *some* tests (test/cql-pytest) against Cassandra, but others (test/alternator) against AWS DynamoDB, and yet others (test/redis) against Redis.
2. Yes, it is indeed a very good idea to run the most recent tests against *different* Scylla versions, running locally or remotely. I did it in the past when asked to test whether a particular version running by a particular client was vulnerable to a known bug for which we had a test. I also do this for bisecting Scylla against a fixed test. I also used it once when I wanted to see what we forgot to backport to a particular branch, and wanted to run the full current test suite of test/alternator against the old version, and see what fails. This was one of the design goals of test/cql-pytest: You don't _have to_ use test/cql-pytest/run (or test.py) - you can run Scylla somewhere manually, and then run "pytest --host ..." and it will test that Scylla installation, which can be any version (and can also be not Scylla ;-)).

**Comment by kostja** [comment]
@nyh yes, you're right about it all, but how important is it? My take is that it's not very important. There is an infinite amount of bugs, so it's important to chase the right ones. That's all I am saying.
